src/hotspot/share/opto/escape.cpp

```diff
@@ -387,6 +398,318 @@ bool ConnectionGraph::compute_escape() {
+  // no profit doing the reduction of inputs.
+  if (!has_noescape_allocate) {
+    NOT_PRODUCT(if (TraceReduceAllocationMerges) tty->print_cr("Will NOT try to reduce Phi %d. There is not any NoEscape Allocate as input.", phi->_idx);)
+    return false;
+  }
+
+  // Try to find a BOT memory Phi coming from same region
+  Node* reg = phi->in(0);
+  for (DUIterator_Fast imax, i = reg->fast_outs(imax); i < imax; i++) {
+    Node* n = reg->fast_out(i);
+    if (n->is_Phi() && n->bottom_type() == Type::MEMORY) {
+      if (_compile->get_alias_index(n->adr_type()) == Compile::AliasIdxBot) {
+        found_memory_edge = true;
+        break;
+      }
+    }
+  }
+
+  if (!found_memory_edge && has_nonnull_nonsr_input) {
+    NOT_PRODUCT(if (TraceReduceAllocationMerges) tty->print_cr("Will NOT try to reduce Phi %d. Did not find memory edge on Region.", phi->_idx);)
+    return false;
+  }
+
+  // Validate outputs:
+  //    Check if we can in fact later replace the uses of the
+  //    current Phi by Phi's of individual fields.
+  //    Conditions checked:
+  //       - The only consumers of the Phi are:
+  //           - AddP (with constant offset)
+  //           -   - Load
+  //           - Safepoint
+  //           - uncommon_trap
+  //           - DecodeN
+  //
+  // TODO: add support for other kind of users.
+  for (DUIterator_Fast imax, i = phi->fast_outs(imax); i < imax; i++) {
+    Node* use = phi->fast_out(i);
+
+    if (use->is_CallStaticJava() || use->Opcode() == Op_SafePoint) {
+      has_call_as_user = true;
+
+      if (use->is_CallStaticJava() && use->as_CallStaticJava()->is_uncommon_trap() == false) {
+        NOT_PRODUCT(if (TraceReduceAllocationMerges) tty->print_cr("Will NOT try to reduce Phi %d. Has Allocate but cannot scalar replace it. CallStaticJava is not a trap.", phi->_idx);)
+        return false;
+      }
+    }
+    else if (use->is_AddP()) {
+      if (use->in(AddPNode::Offset)->find_intptr_t_con(-1) == -1) {
+        NOT_PRODUCT(if (TraceReduceAllocationMerges) tty->print_cr("Will NOT try to reduce Phi %d. Did not find constant input for %d : AddP.", phi->_idx, use->_idx);)
+        return false;
+      }
+
+      for (DUIterator_Fast jmax, j = use->fast_outs(jmax); j < jmax; j++) {
+        Node* use_use = use->fast_out(j);
+
+        if (!use_use->is_Load()) {
+          NOT_PRODUCT(if (TraceReduceAllocationMerges) tty->print_cr("Will NOT try to reduce Phi %d. AddP use is not a Load. %d %s", phi->_idx, use_use->_idx, use_use->Name());)
+          return false;
+        }
+      }
+    }
+    else if (use->is_DecodeN()) {
+      for (DUIterator_Fast jmax, j = use->fast_outs(jmax); j < jmax; j++) {
+        Node* use_use = use->fast_out(j);
+
+        if (!use_use->is_AddP()) {
+          NOT_PRODUCT(if (TraceReduceAllocationMerges) tty->print_cr("Will NOT try to reduce Phi %d. DecodeN use is not a Load. %d %s", phi->_idx, use_use->_idx, use_use->Name());)
+          return false;
+        }
+
```

Comment:
You are missing check for constant offset as you did in previous code. May be move this code into small method.
Also this whole code which checks users could be separate method.